### ISSUE TITLE
torch.angle and torch.atan2 are producing unexpected gradients.

### ISSUE THREAD

**Issue by yuxinyuan** [pytorch/pytorch#98276]
### 🐛 Describe the bug

``torch.angle`` can produce NaN gradients for inputs that are close to (0, 0). ``torch.atan2`` produces NaN gradient when the input is exactly (0, 0)
```python
>>> import torch
>>> a = torch.tensor(5.6843e-14+1.0000e-13j, requires_grad=True)
>>> a
tensor(5.6843e-14+1.0000e-13j, requires_grad=True)
>>> o = torch.angle(a)
>>> o
tensor(1.0539, grad_fn=<AngleBackward0>)
>>> a.grad
>>> o.backward()
>>> a.grad
tensor(nan+nanj)
```
In contrast, ``torch.atan2`` is able to produce the correct gradient.
```python
>>> b = a.detach().clone()
>>> b.requires_grad_(True)
tensor(5.6843e-14+1.0000e-13j, requires_grad=True)
>>> p = torch.atan2(b.imag, b.real)
>>> p
tensor(1.0539, grad_fn=<Atan2Backward0>)
>>> b.grad
>>> p.backward()
>>> b.grad
tensor(-7.5579e+12+4.2962e+12j)
```
However, when the input is exactly (0, 0), ``torch.atan2`` can produce NaN gradients (because of divided by zero issue in the backward function)
```python
>>> c = torch.tensor(0+0j, requires_grad=True)
>>> c
tensor(0.+0.j, requires_grad=True)
>>> torch.angle(c), torch.atan2(c.imag, c.real)
(tensor(0., grad_fn=<AngleBackward0>), tensor(0., grad_fn=<Atan2Backward0>))
>>> c.grad
>>> torch.atan2(c.imag, c.real).backward()
>>> c.grad
tensor(nan+nanj)
```

I found this piece of function in the source code, 
https://github.com/pytorch/pytorch/blob/c263bd43e8e8502d4726643bc6fd046f0130ac0e/torch/csrc/autograd/FunctionsManual.cpp#L493
it turns out that when a complex number is divided by a very small real number, the result will become inf. This is likely a numerical precision issue which causes the NaN gradient of ``torch.angle``.
```python
>>> a
tensor(5.6843e-14+1.0000e-13j, requires_grad=True)
>>> torch.tensor(3+3j) / (a.imag * a.imag + a.real * a.real)
tensor(inf+infj, grad_fn=<DivBackward0>)
```
On the other hand, the backward of ``torch.atan2`` computes the gradient of the real and imag part separately, and there seems to be no numerical precision issue
https://github.com/pytorch/pytorch/blob/c263bd43e8e8502d4726643bc6fd046f0130ac0e/torch/csrc/autograd/FunctionsManual.cpp#L2960

I'm not sure if this should be considered a bug, but maybe ``torch.angle`` and ``torch.atan2`` should have consistent behaviors.
Also, since both ``torch.angle`` and ``torch.atan2`` produce correct forward result, the NaN gradient issue should at least be documented so that users know about this.

### Versions

PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.8.0-63-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 510.54
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @anjali411 @dylanbespalko @mruberry

**Comment by ZailiWang** [comment]
The torch.angle() issue has been fixed in latest version. Would you update to nightly build and confirm?

**Comment by yuxinyuan** [comment]
Yes. The nightly build seems to fix the ``torch.angle`` issue.

**Comment by lezcano** [comment]
`arctan2(0,0) = arctan(0 / 0)`, which is undefined. As such, its gradients are also undefined. See point number 5 in https://pytorch.org/docs/stable/notes/autograd.html#gradients-for-non-differentiable-functions.

Closing as both issues seem either fixed or expected. Feel free to reopen if you think otherwise
